### DISCUSSION TITLE
How do I generate an UnstructuredGrid with polyhedrons?

### DISCUSSION THREAD

**Discussion by PProfizi** [pyvista/pyvista#3414] [Q&A] [answered]
Hello everyone,

I have been trying to generate an UnstructuredGrid using the format `grid = pv.UnstructuredGrid(cells, vtk_cell_type, nodes)`, yet I have not managed to find any information on whether doing this with polygons or polyhedrons is possible.
I have managed to generate a mesh with polyhedrons using VTK and InsertNextCell, yet this requires a loop, so for performance and ease of use I would need to directly initiate a full pv.UnstructuredGrid containing a mix of regular elements and polyhedrons with arrays for connectivities, face connectivities, element types...
I have not found any example showcasing this so I am not sure this is even possible.

Thank you for your help!

**Comment by AlejandroFernandezLuces** [comment, marked as answer]
It is possible, but it is undocumented. 

Polyhedrons are a special case inside VTK, so its format is different from other figures:

![MicrosoftTeams-image](https://user-images.githubusercontent.com/17165476/198543340-e961e862-cf4b-4067-99df-900a40dff8d3.png)

This format is also documented in the [VTK documentation for polyhedron](https://vtk.org/Wiki/VTK/Polyhedron_Support):

> However, previous 3D cells are regular ones whose geometries are fully determined by a set of vertex points with predefined implicit ordering. Unfortunately, such an implicit ordering does not exist in polyhedron cells, the connectivities between polyhedron vertices need to be explicitly defined. The extended vtkUnstructuredGrid stores polyhedron cells as face streams of the following format:

     [numberOfCellFaces, (numberOfPointsOfFace0, pointId0, pointId1, … ), (numberOfPointsOfFace1, pointId0, pointId1, …), … ]. 

I created this small example with the case you propose for you to test. Let me know if this solves your question. I'll open an issue in PyVista to properly document and test this.

```
import pyvista as pv
import numpy as np
from vtk import VTK_POLYHEDRON, VTK_POLYGON


ptIds = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
nodes = np.asarray([[0.01,  0.01,  0.01],
                    [0,  0.01,  0.01],
                    [0,  0.,    0.01],
                    [0.01,  0.,    0.01],
                    [0.01,  0.,    0.],
                    [0,  0.,    0.],
                    [0,  0.01,  0.],
                    [0.01,  0.01,  0.],
                    [0.01, 0.01, 0.02],
                    [0.00, 0.01, 0.02]])

# polyhedron format [NElements, NFaces, Face1NPoints, Face1Point1, Face1Point2..., Face1PointN, FaceNNPoints,...]]
# in polygons and polyhedrons, face order/points order matters
polyhedron = np.asarray([24, 5, 4, 2, 3, 4, 5, 4, 5, 6, 7, 4, 3, 4, 7, 3, 3, 5, 6, 2, 4, 6, 7, 3, 2])
polygon = np.asarray([4, 1, 9, 8, 0])

cells = np.hstack((polyhedron, polygon))
cell_type = [VTK_POLYHEDRON, VTK_POLYGON]
grid = pv.UnstructuredGrid(cells, cell_type, nodes)

plt = pv.Plotter()
plt.add_mesh(grid)
plt.show()


```
